bzt/modules/apachebench.py

```diff
@@ -0,0 +1,204 @@
+        self.start_time = None
+
+    def prepare(self):
+        self.scenario = self.get_scenario()
+        self.tool_path = self._check_installed()
+
+        self.__tsv_file_name = self.engine.create_artifact("apachebench", ".tsv")
+
+        out_file_name = self.engine.create_artifact("apachebench", ".out")
+        self.reader = TSVDataReader(self.__tsv_file_name, self.log)
+        if isinstance(self.engine.aggregator, ConsolidatingAggregator):
+            self.engine.aggregator.add_underling(self.reader)
+
+        self.__out = open(out_file_name, 'w')
+        self.__err = open(self.engine.create_artifact("apachebench", ".err"), 'w')
+
+    def startup(self):
+        args = [self.tool_path]
+        load = self.get_load()
+
+        if load.iterations:
+            args += ['-n', str(load.iterations)]
+        elif load.hold:
+            hold = int(ceil(dehumanize_time(load.hold)))
+            args += ['-t', str(hold)]
+        else:
+            raise ValueError("You must specify either 'iterations' or 'hold-for' for apachebench")
+
+        load_concurrency = load.concurrency
+        args += ['-c', str(load_concurrency)]
+        self.reader.concurrency = load_concurrency
+
+        args += ['-d']  # do not print 'Processed *00 requests' every 100 requests or so
+
+        args += ['-g', str(self.__tsv_file_name)]  # dump stats to TSV file
+
+        # add global scenario headers
+        for key, val in iteritems(self.scenario.get_headers()):
+            args += ['-H', "%s: %s" % (key, val)]
+
+        requests = list(self.scenario.get_requests())
+        if not requests:
+            raise ValueError("You must specify at least one request for apachebench")
+        if len(requests) > 1:
+            self.log.warning("ApacheBench doesn't support multiple requests."
+                             " Only first one will be used.")
+        request = requests[0]
+
+        # add request-specific headers
+        for header in request.headers:
+            for key, val in iteritems(header):
+                args += ['-H', "%s: %s" % (key, val)]
+
+        args += ['-m', request.method]
+
+        if request.timeout is not None:
+            timeout = int(ceil(dehumanize_time(request.timeout)))
+            args += ['-s', str(timeout)]
+
+        args += [request.url]
+
+        self.start_time = time.time()
+        self.process = shell_exec(args, stdout=self.__out, stderr=self.__err)
+
+    def check(self):
+        ret_code = self.process.poll()
+        if ret_code is None:
+            return False
+        self.log.info("ApacheBench tool exit code: %s", ret_code)
+        if ret_code != 0:
+            raise RuntimeError("ApacheBench tool exited with non-zero code")
+        return True
+
+    def shutdown(self):
+        """
+        If tool is still running - let's stop it.
+        """
+        shutdown_process(self.process, self.log)
+        if self.__out and not self.__out.closed:
+            self.__out.close()
```

Comment:
Handles opened on prepare stage must be closed on post-process stage. Because there is a chance to not have shutdown call at all.